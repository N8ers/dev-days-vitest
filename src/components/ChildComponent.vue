<template>
  <div class="container">
    <h3>Child</h3>
    <p>{{ props.message }}</p>
    <hr />
    <input v-model="childMessage" type="text" />
    <button @click="emitToParent">emit message to parent!</button>
  </div>
</template>

<script setup>
import { ref } from "vue";

const props = defineProps({
  message: {
    type: String,
    required: false,
    default: "I am the default message",
  },
});

const emit = defineEmits(["emitMessageUp"]);

const childMessage = ref("");

const emitToParent = () => {
  emit("emitMessageUp", childMessage.value);
};
</script>

<style scoped>
.container {
  margin: 20px;
  border: 1px solid black;
  padding: 10px;
  width: 50%;
}
</style>
