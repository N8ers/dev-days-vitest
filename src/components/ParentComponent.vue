<template>
  <div>
    <ChildComponent
      message="I am a message from the parent!"
      @emitMessageUp="handleEmitMessageUp"
    />

    <div v-if="childComponentMessage.length">
      Child Compoent Message: {{ childComponentMessage }}
    </div>
  </div>
</template>

<script setup>
import { ref } from "vue";
import ChildComponent from "./ChildComponent.vue";

const childComponentMessage = ref("");

const handleEmitMessageUp = (message) => {
  childComponentMessage.value = message;
};
</script>
