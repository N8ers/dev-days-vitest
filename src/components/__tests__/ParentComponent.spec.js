import { mount } from "@vue/test-utils";
import { describe, expect, test } from "vitest";

import ParentComponent from "@/components/ParentComponent.vue";
import ChildComponent from "@/components/ChildComponent.vue";

describe("ChildComponent.vue", () => {
  describe("Renders content on load", () => {
    test("Renders default content", () => {
      const wrapper = mount(ParentComponent);

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe("Component conditionally renders 'childComponentMessage'", () => {
    test("Triggering emit from ChildComponent renders 'childComponentMessage'", async () => {
      const wrapper = mount(ParentComponent);
      // content should be hidden until emit
      expect(wrapper.find('[data-test="childComponentMessage"]').exists()).toBe(
        false
      );
      // default message is empty
      expect(wrapper.vm.childComponentMessage).toBe("");
      const childComponent = wrapper.findComponent(ChildComponent);
      // simulate an emit from ChildComponent
      await childComponent.vm.$emit("emitMessageUp", "a chill message");
      // check that the emit was received
      expect(wrapper.vm.childComponentMessage).toBe("a chill message");
      // check that the template renders `v-if`
      // await flushPromises();
      const childComponentMessage = wrapper.find(
        '[data-test="childComponentMessage"]'
      );
      expect(childComponentMessage.exists()).toBe(true);
      expect(childComponentMessage.text()).toBe(
        "Child Compoent Message: a chill message"
      );
    });
  });
});
