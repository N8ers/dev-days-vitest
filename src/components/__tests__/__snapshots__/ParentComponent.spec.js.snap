// Vitest Snapshot v1

exports[`ChildComponent.vue > Renders content on load > Renders default content 1`] = `
<div class="container" data-v-3e4aa81a="">
  <h3 data-v-3e4aa81a="">Parent</h3>
  <!--v-if-->
  <div class="container" data-v-71c74cf1="" data-v-3e4aa81a="">
    <h3 data-v-71c74cf1="">Child</h3>
    <p data-v-71c74cf1="">I am a message from the parent!</p>
    <hr data-v-71c74cf1=""><input type="text" data-v-71c74cf1=""><button data-v-71c74cf1="">emit message to parent!</button>
  </div>
</div>
`;
